Added option to retrieve the colors used in the QR code from the image.

david476 opened the pull request.

cozmo commented:
Awesome work on the PR. It's been a pretty busy week for me, but I'll give it a look soon. 

Meanwhile in #77 I've added the options object for another option we needed to add, so you may be able to leverage some of the work there.

cozmo commented:
If you want faster iteration on this lint stuff you can just run `npm run-script lint` locally.

david476 commented:
@cozmo It looks like the tests are failing due to how the objects are getting serialized? Not sure how to fix that...

david476 commented:
Yeah, I'm not sure how to fix that. it seems to be caused by data type issues...

cozmo commented on src/color-retriever/index.ts:
```diff
@@ -0,0 +1,188 @@
+color space for averaging (with no regard for alpha), and then converted back to RGB. Alpha values are simply averaged
+directly. */
+export function retrieveColors(location: QRLocation,
+                               extracted: {matrix: BitMatrix; mappingFunction: (x: number, y: number) => Point; },
```
Instead of defining this here, might make sense to define an `Extracted` type in extractor and then reference it here, I feel like that would be a little cleaner per se.

cozmo commented on src/index.ts:
```diff
@@ -21,21 +22,22 @@ export interface QRCode {
 }
 
-function scan(matrix: BitMatrix): QRCode | null {
+function scan(matrix: BitMatrix, sourceData: Uint8ClampedArray, sourceWidth: number, scanOptions: Options): QRCode | null {
```
Not sure if this makes sense or not, but the width and the height are already available on the binarized bitMatrix. Not sure we need to pass them in again.

david476 commented:
Those all sound good, I'll update all that soon. Changing the `(QR)Color` interface should fix the tests but, to confirm, you're OK with the way I implemented them?

david476 commented:
For testing I added some checks for the color space transformations and effectively `end-to-end-tests.ts` but that only looks at color. For populating the test data I just added on to `generate-test-data.ts` but that might want to be broken out into a separate script.

cozmo commented on src/color-retriever/index.ts:
```diff
@@ -0,0 +1,194 @@
+}
+
+interface CIELabColor {
+  L: number;
```
Is it expected that the `L` is capitalized but the other attribute aren't?

david476 replied:
Most of the [reference material](https://en.wikipedia.org/wiki/CIELAB_color_space) I've seen *generally* capitalizes them that way... Up to you though, obviously it doesn't follow javascript conventions.

cozmo commented on src/color-retriever/index.ts:
```diff
@@ -0,0 +1,194 @@
+
+// Color space conversions from http://www.easyrgb.com/en/math.php
+
+// Converts an RGB color ([r, g, b] or [r, g, b, a] - a is ignored) to CIELab ([L*, a*, b*]).
```
This comment isn't strictly correct anymore I don't think - Can probably just remove the array descriptions from it?

cozmo commented on src/color-retriever/index.ts:
```diff
@@ -0,0 +1,194 @@
+  return {L: l, a, b};
+}
+
+// Converts a CIELab color ([L*, a*, b*] - ignores additional values) to RGB ([r, g, b]).
```
Same re: this comment no longer needing the array descriptions.

cozmo commented on tests/generate-test-data.ts:
```diff
@@ -25,4 +25,15 @@ const report = {
+    const output = jsQR(imageData.data, imageData.width, imageData.height, {retrieveColors: true});
+
+    await fs.writeFile(path.join("src", "color-retriever", "test-data", t, "output.json"), JSON.stringify(output.colors, null, 2), "utf8");
+  }
```
I actually don't think we strictly need this, since I think the test cases are sort of...stand alone. Now that we've created them I would expect them to not break as we change the library? If they do we can re-visit needing to autogen them, but for now would be fine leaving them standalone?

david476 replied:
Seems useful to leave for adding more color tests but splitting it out makes sense. I'll make it a separate file in the color retrieving folder?

cozmo commented on src/color-retriever/test.ts:
```diff
@@ -0,0 +1,47 @@
+function roundComponents(color) {
+  for(var component in color) {
+    color[component] = Math.round(color[component])
+  }
```
This function doesn't return anything so I don't actually think the tests are doing anything, since they're always comparing `undefined` to `undefined`

david476 replied:
I think it should work (given the mutability of arrays) but definitely not intentional :)

cozmo replied:
The mutability doesn't matter here I don't think. This function has to return, and so as such returns undefined. We use it in the assertions above, so they all become `expect(undefined).toBe(undefined)`. Unless I'm missing something?

david476 replied:
Oh boy, I must have been reeeealy out of it. NVM

cozmo commented:
Just left one final comment since I tried to merge this. but realized the conversions tests weren't actually doing anything. Right now there's a merge conflict on master, but feel free to ignore that, I'll handle it since I caused it. That said not really sure how best to fix these tests so would rather defer to you.

david476 commented:
@cozmo Sorry for disappearing for so long! It's been a busy school year... If I finish up this pull request (I think I had more to do even though you approved it...), still up for merging it?
